### PR TITLE
Implements directory based authentication services

### LINKED ISSUE #4: Implement AD and LDAP auth services
Base those on the code and expertise collected in WebAPI

### PR DESCRIPTION
**Pull request by wivern**
Resolves #4

### CODE REVIEW

**Comment by wivern** [comment]
> 1. In WebAPI we have both `security.ad.searchBase` and `security.ad.searchFilter`. I do see only `searchFilter` here

I believe two ldap queries is redundant. In WebAPI there was an issues when trying to restrict access based on AD groups.
But as `AdAuthenticator` uses `LdapTemplate.authenticate` method it's possible to write a complex ldap filter is a single statement, like this one: `(&(userPrincipalName={0})(memberOf=CN=Domain Users,CN=Users,DC=odysseusinc,DC=domain))`

**Comment by pavgra** [comment]
But WebAPI's ADRealm also uses the `ldapTemplate` so why then we had a need in two params there?
```
@Override
    protected AuthenticationInfo queryForAuthenticationInfo(AuthenticationToken token, LdapContextFactory ldapContextFactory) throws NamingException {

        if (Objects.nonNull(ldapTemplate) && StringUtils.isNotBlank(searchFilter) && StringUtils.isNotBlank(searchString)) {
            UsernamePasswordToken upToken = (UsernamePasswordToken) token;
            String userPrincipalName = getUserPrincipalName(upToken.getUsername());

            String userSearch = String.format(searchString, userPrincipalName);
            List<UserPrincipal> result = ldapTemplate.search("", userSearch, SearchControls.SUBTREE_SCOPE,
                    userMapper);

            if (result.size() == 1) {
                UserPrincipal userPrincipal = result.iterator().next();
                List<String> filterResult = ldapTemplate.search("", String.format(searchFilter, userPrincipal.getUsername()),
                        SearchControls.SUBTREE_SCOPE, dnAttributesMapper);
                if (!filterResult.isEmpty()) {
                    LdapContext ctx = null;
                    try {
                        ctx = ldapContextFactory.getLdapContext(upToken.getUsername(), String.valueOf(upToken.getPassword()));
                    } finally {
                        LdapUtils.closeContext(ctx);
                    }

                    return new SimpleAuthenticationInfo(userPrincipal, upToken.getPassword(), getName());
                }
            } else {
                LOGGER.warn("Multiple results found for {}", userPrincipalName);
            }
        } else {
            return super.queryForAuthenticationInfo(token, ldapContextFactory);
        }
        return null;
    }
```

**Comment by wivern** [comment]
Because WebAPI implementation uses `LdapTemplate.search` but authenticator `LdapTemplate.authenticate` which calls `search` under the hood.
Unfortunately cannot remember exactly why not implemented `authenticate` but I guess since it was not work as expected. Currently explicit search could be fully replaced with authenticate.

File: README.md [hunk 1]
```diff
@@ -106,6 +106,73 @@ authenticator:
...
+`searchFilter` is a formatted LDAP query using to authenticate user. Query should include at least parameter for
+    user who is logging in. Formatted parameters should be encoded like for `MessageFormat.format`. 
+    Query example: `uid={0}`. To succesful authentication `searchFilter` should return exactly one record,
+    if successful authentication service would try to bind to directory using distingueshed name of result node and
+    provided password. 
+
+`url` defines server URL including protocol and port, e.g. `ldap://localhost:389`
+`baseDn` Distiguesh name of system account having permissions to search across directory
+`password` Password of account represented by `baseDn` above
+`baseDn` Distinguesh name of top level tree node that gives subtree to search into
```
Comment (pavgra) [inline README.md]: The same parameter name as on line 119

File: README.md [hunk 2]
```diff
@@ -106,6 +106,73 @@ authenticator:
           firstName: name.split(' ')[0]
           lastName: name.split(' ')[1]
 ```
+### LDAP authentication
+
+Authentication across directory requires special system account that used to search across directory.
+`searchFilter` is a formatted LDAP query using to authenticate user. Query should include at least parameter for
```
Comment (pavgra) [inline README.md]: `using to` -> `used to`
`at least parameter` -> `at least a parameter`

File: README.md [hunk 3]
```diff
@@ -106,6 +106,73 @@ authenticator:
...
+    user who is logging in. Formatted parameters should be encoded like for `MessageFormat.format`.
```
Comment (pavgra) [inline README.md]: > Formatted parameters should be encoded like for `MessageFormat.format`. 

Honestly, not sure whether I get it

File: README.md [hunk 4]
```diff
@@ -106,6 +106,73 @@ authenticator:
           firstName: name.split(' ')[0]
           lastName: name.split(' ')[1]
 ```
+### LDAP authentication
+
+Authentication across directory requires special system account that used to search across directory.
+`searchFilter` is a formatted LDAP query using to authenticate user. Query should include at least parameter for
+    user who is logging in. Formatted parameters should be encoded like for `MessageFormat.format`. 
+    Query example: `uid={0}`. To succesful authentication `searchFilter` should return exactly one record,
```
Comment (pavgra) [inline README.md]: `To succesful authentication` -> `For successful authentication`

File: README.md [hunk 5]
```diff
@@ -106,6 +106,73 @@ authenticator:
...
+    if successful authentication service would try to bind to directory using distingueshed name of result node and
```
Comment (pavgra) [inline README.md]: Is this the second part of a single sentence?

File: README.md [hunk 6]
```diff
@@ -106,6 +106,73 @@ authenticator:
...
+        fieldsToExtract:
+          firstName: displayName
+          lastName: sn
+``` 
+
+### Active Directory authentication
+
+Active Directory mostly is similar to LDAP authentication, except:
+- `baseDn` allows to provide username istead of distingueshed name 
+- `domainSuffix` AD domain name, when is not set `baseDn` should be FQDN, not just login.
```
Comment (pavgra) [inline README.md]: > AD domain name, when is not set `baseDn` should be FQDN, not just login.

I would split into two sentences, something like this:
> defines AD domain name. When it is not set, the `baseDn` should have format of FQDN (not just login).

File: README.md [hunk 7]
```diff
@@ -106,6 +106,73 @@ authenticator:
...
+
+Authentication across directory requires special system account that is used to conduct search.
+`searchFilter` is a formatted LDAP query used to authenticate user. Query should include at least a parameter for
+    user who is logging in. Formatted parameters should be encoded to fit [`MessageFormat.format`](https://docs.oracle.com/javase/8/docs/api/java/text/MessageFormat.html]). 
+    Query example: `uid={0}`. For succesful authentication `searchFilter` must return exactly one record.
+    When a single record is found, authentication service will try to bind to directory using distingueshed name of result node and
+    provided password. 
+
+`url` defines server URL including protocol and port, e.g. `ldap://localhost:389`
+`userDn` Distigueshed name of system account having permissions to search across directory
```
Comment (pavgra) [inline README.md]: Typo. DistiNguished

File: src/test/resources/application-testad.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    embedded:
+      base-dn: DC=example,DC=com
+      credential:
+        username: CN=John Smith,CN=Users,DC=example,DC=com
+        password: password
+      ldif: classpath:ad-users.ldif
+      port: 9389
+      validation:
+        enabled: true
+authenticator:
```
Comment (pavgra) [inline src/test/resources/application-testad.yml]: Why this method config is put separately while all others are in `src/test/resources/application.yml`? I would keep things consistent

File: src/test/resources/application.yml [hunk 9]
```diff
@@ -68,9 +68,32 @@ authenticator:
...
+          firstName: displayName
+          lastName: sn
 security:
   jwt:
     token:
       secretKey: 129DF19C8A91AFD8375A2826A33539K01ACQ778QOJFAA9MGWLWH73PLXVFVHBR7860MTIE2O8EEVF9KCO77P6A7NUNX4XHAGCRFSBWG879XPDOIN6C2LFCKJI002OIABS4D6Q9VMJJIX8UCE48EF
-      validityInSeconds: 60
\ No newline at end of file
+      validityInSeconds: 60
+spring:
```
Comment (pavgra) [inline src/test/resources/application.yml]: Again, to keep things consistent, why not to put it into separate `application-test-ldap.yml` as with AD?

**Pull request merged**